### PR TITLE
Issue 5086

### PR DESCRIPTION
**Pull request by phadej**
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: Cabal/Distribution/PackageDescription/Check.hs [hunk 1]
```diff
@@ -1664,6 +1665,34 @@ checkUnicodeXFields gpd
...
+            ]
+    | otherwise = []
+  where
+    mn = autogenPathsModuleName pd
+
+    checkLib :: Library -> Bool
+    checkLib l = mn `elem` exposedModules l && checkExts (l ^. L.defaultExtensions)
+
+    checkBI :: BuildInfo -> Bool
+    checkBI bi = mn `elem` otherModules bi && checkExts (bi ^. L.defaultExtensions)
```
Comment (23Skidoo) [inline Cabal/Distribution/PackageDescription/Check.hs]: Should we also check `autogen-modules`?

File: Cabal/Distribution/Types/PackageDescription.hs [hunk 2]
```diff
@@ -369,22 +369,15 @@ withForeignLib pkg_descr f =
 -- dependencies.
 allBuildInfo :: PackageDescription -> [BuildInfo]
 allBuildInfo pkg_descr = [ bi | lib <- allLibraries pkg_descr
```
Comment (23Skidoo) [inline Cabal/Distribution/Types/PackageDescription.hs]: There should be a mention in the changelog that the semantics of this function changed. Also the Haddock comment should be updated.

**Comment by 23Skidoo** [comment]
Looks like the test needs to be fixed.

**Comment by phadej** [comment]
I have to use `cabal-version: 2.2`. Good. Will fix ASAP

File: Cabal/Distribution/PackageDescription/Check.hs (hunk 1 above)
Comment (phadej) [reply Cabal/Distribution/PackageDescription/Check.hs]: I guess we can

**Comment by 23Skidoo** [comment]
There's still a test failure on AppVeyor.

File: Cabal/Distribution/Types/PackageDescription.hs [hunk 3]
```diff
@@ -364,27 +364,21 @@ withForeignLib pkg_descr f =
 -- ---------------------------------------------------------------------------
 -- The BuildInfo type
 
--- | The 'BuildInfo' for the library (if there is one and it's buildable), and
--- all buildable executables, test suites and benchmarks.  Useful for gathering
--- dependencies.
+-- | All 'BuildInfo' in the 'PackageDescription':
+-- libraries, executables, test-suites and benchmarks.
```
Comment (23Skidoo) [inline Cabal/Distribution/Types/PackageDescription.hs]: You forgot foreign libraries :-).

Comment (phadej) [reply Cabal/Distribution/Types/PackageDescription.hs]: *libraries*

**Comment by phadej** [comment]
I hardcoded `cabalVersion` to be 2.2 (part of #5094). I'll make a sweep to remove 2.1 and 1.25 occurrences from the code.

**Comment by 23Skidoo** [comment]
Still seems to fail, plus there's now a merge conflict. I'm not force-pushing to your branch because the PR gets closed when a maintainer does that.

**Comment by phadej** [comment]
the problem with `CustomPlain` test is that it fails locally for me but for different reason:

```
ghc: can't find a package database at /home/ogre/cabal/cabal-testsuite/PackageTests/CustomPlain/cabal.dist/work/./dist/packagedb/ghc-8.2.2
```

**Comment by 23Skidoo** [comment]
Closing in favour of #5097.

**Pull request closed without merging**